### ISSUE TITLE
MapEntryStages.innerDefaultReplaceValue()

### ISSUE THREAD

**Issue by leventov** [OpenHFT/Chronicle-Map#169]
Regarding this commit: https://github.com/OpenHFT/Chronicle-Map/commit/79c7701a52058361150f7d2e7139cbd0d4211a9f, @RobAustin there was a reason why the code was written the way it was - it's that during relocation a different value alignment might be required, therefore a different number of chunks might be needed.

**Comment by RobAustin** [comment]
The original code was causing a newly added test  to fail. These tests now pass. If you feel there is a mistake please add a falling test.

**Comment by leventov** [comment]
```java
  @Test
  public void testValueAlignmentRelocation() {
    ChronicleMap<byte[], byte[]> map = ChronicleMapBuilder
        .of(byte[].class, byte[].class)
        .averageKeySize(5)
        .averageValueSize(5)
        .keySizeMarshaller(SizeMarshaller.stopBit())
        .valueSizeMarshaller(SizeMarshaller.stopBit())
        .entryAndValueOffsetAlignment(8)
        .actualSegments(1)
        .actualChunkSize(2)
        .entries(10)
        .create();
    Random r = new Random(0);
    for (int firstKeySize = 1; firstKeySize < 10; firstKeySize++) {
      byte[] firstKey = new byte[firstKeySize];
      byte[] firstValue = new byte[8];
      r.nextBytes(firstKey);
      r.nextBytes(firstValue);
      for (int secondKeySize = 1; secondKeySize < 10; secondKeySize++) {
        byte[] secondKey = new byte[secondKeySize];
        r.nextBytes(secondKey);
        byte[] secondValue = new byte[1];
        r.nextBytes(secondValue);
        map.clear();
        map.put(firstKey, firstValue);
        map.put(secondKey, secondValue);
        byte[] thirdValue = new byte[16];
        r.nextBytes(thirdValue);
        map.put(firstKey, thirdValue);
        for (int i = 0; i < 10; i++) {
          map.put(new byte[] {(byte) i}, new byte[] {(byte) i});
        }
        Assert.assertTrue(Arrays.equals(map.get(firstKey), thirdValue));
      }
    }
  }
```

**Comment by leventov** [comment]
Also I found another problem - apparently `ChronicleMapBuilder.entryAndValueOffsetAlignment()` is not respected for _entries_, only for values. Seems like too obvious mistake, but I don't see why it's not

**Comment by RobAustin** [comment]
Thanks

**Comment by RobAustin** [comment]
I've added this test above as net.openhft.chronicle.map.ValueAlignmentRelocation#testValueAlignmentRelocation

**Comment by dpisklov** [comment]
@RobAustin @leventov the test provided passes for me and on teamcity. Does it fail for anyone? If not - what do I need to fix (apart from `ChronicleMapBuilder.entryAndValueOffsetAlignment()` issue)?

**Comment by dpisklov** [comment]
Actually I see this code:
```
public final long entrySize(long keySize, long valueSize) {
        long sizeOfEverythingBeforeValue = sizeOfEverythingBeforeValue(keySize, valueSize);
        return innerEntrySize(sizeOfEverythingBeforeValue, valueSize);
    }

    public long innerEntrySize(long sizeOfEverythingBeforeValue, long valueSize) {
        if (!mh.m().constantlySizedEntry && mh.m().couldNotDetermineAlignmentBeforeAllocation)
            sizeOfEverythingBeforeValue += mh.m().worstAlignment;
        int alignment = mh.m().alignment;
        return alignAddr(sizeOfEverythingBeforeValue, alignment) +
                alignAddr(valueSize, alignment);
    }
```
the `        int alignment = mh.m().alignment;` part is using the alignment from `ChronicleMapBuilder.entryAndValueOffsetAlignment()` and is used to align entry - is it not what we need?

**Comment by leventov** [comment]
@dpisklov I didn't test it inside the project. The test the I provided fails for me with `3.17.0` Chronincle Map dependency. Doesn't it fail in CI?

**Comment by dpisklov** [comment]
@leventov as I mentioned, it passes both on my machine and teamcity

**Comment by leventov** [comment]
I don't understand how this is possible. I double checked, on my machine this test passes on 3.16.4 and fails 3.17.0. The reason is https://github.com/OpenHFT/Chronicle-Map/commit/79c7701a52058361150f7d2e7139cbd0d4211a9f.

**Comment by dpisklov** [comment]
Are you using Windows by a chance?..

**Comment by leventov** [comment]
Mac OS

**Comment by dpisklov** [comment]
Wonder if it's OS-specific. My machine and teamcity are Linux...
@RobAustin does test fail on your mac?

**Comment by RobAustin** [comment]
I'll run it on mac and let you know.

**Comment by RobAustin** [comment]
I agree it fails on the mac, it always fails when "firstKeySize=5, second key=1"

**Comment by RobAustin** [comment]
I have further simplified the failing test above and removed the randomness out of it. This is so that we can focus on fixing a reproducible example, with the same data each time, see below:

````

    @Test
    public void testValueAlignmentRelocationNoRandomTest() {
        ChronicleMap<byte[], byte[]> map = ChronicleMapBuilder
                .of(byte[].class, byte[].class)
                .averageKeySize(5)
                .averageValueSize(5)
                .keySizeMarshaller(SizeMarshaller.stopBit())
                .valueSizeMarshaller(SizeMarshaller.stopBit())
                .entryAndValueOffsetAlignment(8)
                .actualSegments(1)
                .actualChunkSize(2)
                .entries(10)
                .create();

        byte[] firstKey = "austi".getBytes(ISO_8859_1);
        byte[] firstValue = "12345678".getBytes(ISO_8859_1);

        byte[] secondKey = "h".getBytes(ISO_8859_1);
        byte[] secondValue = "a".getBytes(ISO_8859_1);

        map.put(firstKey, firstValue);
        map.put(secondKey, secondValue);

        byte[] thirdValue = "1234567890123456".getBytes(ISO_8859_1);
        map.put(firstKey, thirdValue);
        map.put(("Hello").getBytes(ISO_8859_1), "world".getBytes(ISO_8859_1));

        String actual = toString(map.get(firstKey));
        String expected = toString(thirdValue);

        Assert.assertEquals(expected,actual);
    }
````

It currently fails with the following issue [when run on the mac ]:

org.junit.ComparisonFailure: 
Expected :1234567890123456
Actual   :123456789012Hel

**Comment by RobAustin** [comment]
@leventov When we use .createPersistedTo(<file>); rather than .create(); this issue is not reproducible, so it would appear that it is related to how the data is persisted, do you have any explanation for this. We would have expected the data to be persisted in the same way regardless of if it was stored to a file or not.

**Comment by RobAustin** [comment]
@leventov Also, this test is reproducible on Linux, it was not being run, because the surefire plugin expects that all test classes end in "Test", hence the test was being skipped in our CI environment and when run; it was failing on Linux just like it fails on the mac.

**Comment by leventov** [comment]
@RobAustin could you please test if your simplified test fails with 79c7701a52058361150f7d2e7139cbd0d4211a9f commit reverted?

I may not have time to research this question very soon, but when I was on it I was sure that 79c7701a52058361150f7d2e7139cbd0d4211a9f is wrong. I don't argue that there is some bug that 79c7701a52058361150f7d2e7139cbd0d4211a9f appears to fix (or mask), but it's buggy itself too.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Map-3.17.2](https://github.com/OpenHFT/Chronicle-Map/releases/chronicle-map-3.17.2), [BOM-2.17.244](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.244)
